### PR TITLE
Add support for Tracy 0.9 as a tracing backend.

### PR DESCRIPTION
**Pull request by jamii**
* Add a new tracer backend for https://github.com/wolfpld/tracy. The client is only linked when tb is built with `-Dtracer-backend=tracy`.
* If the tracy backend is enabled, divert logging and allocation statistics to tracy.
* Build the tracer backends in CI, to prevent accidental breakage.

https://user-images.githubusercontent.com/340884/204653140-da12df47-bd1a-481e-9715-d25428ab5b3d.mp4

### CODE REVIEW

File: src/tracer.zig [hunk 1]
```diff
@@ -93,227 +83,424 @@ pub const Event = union(enum) {
...
+        event_group: EventGroup,
+        event: Event,
+        src: std.builtin.SourceLocation,
+    ) void {
+        // The event must not already have been started.
+        assert(slot.* == null);
+        c.___tracy_fiber_enter(event_group.name());
+        const name = std.fmt.bufPrint(&print_buffer, "{}", .{event}) catch name: {
+            const dots = "...";
+            std.mem.copy(u8, print_buffer[print_buffer.len - dots.len ..], dots);
```
Comment (sentientwaffle) [inline src/tracer.zig]: Can you swap all the `mem.copy`s with `util.copy_{disjoint,left,right}` (as appropriate)?

**Comment by jamii** [comment]
I think I'm going to have to turn https://github.com/tigerbeetledb/tigerbeetle/blob/40b022c54a2363387091e19d75abe095280c0e15//src/clients/node/package.json#L43 into a `zig build` command :(

Or else figure out why tb_client is even referring to tracer anyway.

File: src/tracer.zig [hunk 2]
```diff
@@ -93,227 +83,424 @@ pub const Event = union(enum) {
...
+
+    pub fn flush() void {}
+
+    pub fn log_fn(
+        comptime level: std.log.Level,
+        comptime scope: @TypeOf(.EnumLiteral),
+        comptime format: []const u8,
+        args: anytype,
+    ) void {
+        const level_txt = comptime level.asText();
```
Comment (sentientwaffle) [inline src/tracer.zig]: Nit: spell out "text" (or maybe "string") in the suffix.

**Comment by sentientwaffle** [comment]
It could just be the `config.zig` change getting pulled in. Or are you referring to https://github.com/tigerbeetledb/tigerbeetle/actions/runs/3578247700/jobs/6018182393?

(For the node client in particular though, the version check seems to trigger more often than I would expect).

**Comment by jamii** [comment]
The node build is not going through `zig build` at all so it doesn't have `tigerbeetle_build_options` generated. It's not totally obvious to me why it refers anything with tracing though - the entire callstack in the error is in the lsm which the client shouldn't care about. Maybe it's because I'm doing `usingnamespace` in tracer now instead of `const Tracer = ...`? It feels the same semantically, but maybe has different laziness in the compiler.

As for the go build, `zig build tb_client -Drelease-safe -Dtarget=x86_64-windows` works for me locally. Not sure why the workflow fails.

File: build.zig [hunk 3]
```diff
@@ -5,6 +5,8 @@ const TracerBackend = enum {
     none,
     // Writes to a file (./tracer.json) which can be uploaded to https://ui.perfetto.dev/
     perfetto,
+    // Sends data to https://github.com/wolfpld/tracy.
+    tracy,
```
Comment (sentientwaffle) [inline build.zig:9]: The "Which backend to use for tracing" needs to be updated to include tracy. Or the "One of: ..." part can just be removed; I think `zig build -h` automatically lists the possible options for enums.

**Comment by sentientwaffle** [comment]
The node client is going to be updated to use tb_client (and this `build.zig`) in the next few weeks anyway. You could do a hacky workaround in the mean time by checking e.g. `@import("root")` before checking `tigerbeetle_build_options`. :thinking:

File: .gitmodules [hunk 4]
```diff
@@ -1,3 +1,6 @@
 [submodule "src/clients/java/src/zig/lib/jui"]
 	path = src/clients/java/src/zig/lib/jui
 	url = https://github.com/zig-java/jui.git
+[submodule "deps/tracy"]
```
Comment (sentientwaffle) [inline .gitmodules]:
```suggestion
[submodule "tools/tracy"]
```

File: src/tracer.zig (hunk 2 above)
Comment (jamii) [reply src/tracer.zig]: (I copy-pasted that from zig stdlib :grin:)

File: src/tracer.zig [hunk 5]
```diff
@@ -423,15 +424,15 @@ const TracerTracy = struct {
...
         const prefix = if (scope == .default) ": " else "(" ++ @tagName(scope) ++ "): ";
         const message = std.fmt.bufPrint(
             &print_buffer,
-            level_txt ++ prefix ++ format,
+            level_text ++ prefix ++ format,
             args,
         ) catch message: {
             const dots = "...";
-            std.mem.copy(u8, print_buffer[print_buffer.len - dots.len ..], dots);
+            util.copy_disjoint(u8, print_buffer[print_buffer.len - dots.len ..], dots);
```
Comment (sentientwaffle) [inline src/tracer.zig]: You also need to pass either `.exact` or `.inexact` as the first argument (these are exact I think).

Comment (jamii) [reply src/tracer.zig]: Uh, getting bitten by the fact that the backend code is not type-checked if not enabled.

**Comment by jamii** [comment]
The builds that are failing work for me when run locally in a freshly checked out repo. I'm at a loss.

**Comment by jamii** [comment]
Looks like we're green apart from the node version check.

**Review by sentientwaffle — APPROVED** [review]

**Pull request merged**